File: PDOO/Napakalaki-Ruby/lib/player.rb
#encoding: UTF-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module NapakalakiGame

  require_relative "treasure.rb"
  require_relative "dice.rb"
  require_relative "bad_consequence.rb"
  require_relative "card_dealer.rb"

  class Player
    @@MAXLEVEL=10
    def initialize(name)
      @name=name
      @level=1
      @dead=true
      @canISteal=true
      @enemy=nil
      @hiddenTreasures=Array.new
      @visibleTreasures=Array.new
      @pendingBadConsequence=nil
    end
    
    def copyPlayer(p)
      @name = p.name
      @level=p.level
      @dead=p.dead
      @canISteal=p.canISteal
      @enemy=p.enemy
      @hiddenTreasures=p.hiddenTreasures
      @visibleTreasures=p.visibleTreasures
      @pendingBadConsequence=p.pendingBadConsequence     
    end
    
    def self.MaxLevel
      return @@MAXLEVEL
    end
    
    protected
    
    def enemy
      return @enemy
    end
    
    def getHiddenTreasures
      @hiddenTreasures
    end
  
    def getVisibleTreasures
      @visibleTreasures    
    end
    
    public
  
    def getName
      return @name
    end
  
    def isDead
      return @dead
    end
  
    def combat(m) 
      myLevel = getCombatLevel()
      monsterLevel = getOponentLevel(m)
      if(myLevel > monsterLevel)
        applyPrize(m)
        if(@level >= @@MAXLEVEL)
          combatResult = NapakalakiGame::CombatResult::WINGAME
        else
          combatResult = NapakalakiGame::CombatResult::WIN
        end
      else
        applyBadConsequence(m)
        sc = shouldConvert
        if(sc)
          combatResult = NapakalakiGame::CombatResult::LOSEANDCONVERT
        else
          combatResult = NapakalakiGame::CombatResult::LOSE
        end
      end
      return combatResult
    end
  
    def makeTreasureVisible(t) 
      canI = canMakeTreasureVisible(t)
      if(canI)
        @visibleTreasures << t
        @hiddenTreasures.delete_at(@hiddenTreasures.index(t))
      end
    
    end
  
    def discardVisibleTreasure(t) 
      @visibleTreasures.delete_at(@visibleTreasures.index(t))
      if(@pendingBadConsequence != nil && !@pendingBadConsequence.isEmpty)
        @pendingBadConsequence.substractVisibleTreasure(t)
      end
      dieIfNoTreasures
    end
  
    def discardHiddenTreasure(t)
      @hiddenTreasures.delete_at(@hiddenTreasures.index(t))
      if(@pendingBadConsequence != nil && !@pendingBadConsequence.isEmpty)
        @pendingBadConsequence.substractHiddenTreasure(t)
      end
      dieIfNoTreasures
    end
  
    def validState 
      if((@pendingBadConsequence == nil || @pendingBadConsequence.isEmpty) && @hiddenTreasures.size <= 4)
        state=true
      else
        state=false
      end
      return state
    end
  
    def initTreasures 
      dealer = CardDealer.getInstance
      dice = Dice.instance
      bringToLife
      treasure = dealer.nextTreasure
      @hiddenTreasures << treasure
      number = dice.nextNumber
      
      if(number > 1)
        treasure = dealer.nextTreasure
        @hiddenTreasures << treasure
      end
      
      if(number == 6)
        treasure = dealer.nextTreasure
        @hiddenTreasures << treasure
      end
    end

    def getLevels
      return @level
    end
  
    def stealTreasure() 
      treasure = nil
      canI = self.canISteal
      if(canI)
        canYou = @enemy.canYouGiveMeATreasure
        if(canYou)
          treasure = @enemy.giveMeATreasure
          @hiddenTreasures << treasure
          haveStolen
        end
      end
      return treasure
    end
  
    def setEnemy(enemy) 
      @enemy=enemy
    end
 
    def canISteal 
      return @canISteal
    end
  
    def discardAllTreasures
      auxV = Array.new(@visibleTreasures)
      auxV.each { |v|
        self.discardVisibleTreasure(v)
      }
      
      auxH = Array.new(@hiddenTreasures)
      auxH.each { |h| 
        self.discardHiddenTreasure(h)
      }
    end
    
    def canYouGiveMeATreasure 
      if(@hiddenTreasures.size > 0)
        puede=true
      else
        puede=false
      end
      return puede
    end
    
    def giveMeATreasure
      if(canYouGiveMeATreasure)
        indice=rand(@hiddenTreasures.size)+1
        tesoro=@hiddenTreasures.at(indice)
        @hiddenTreasures.delete_at(indice)
        return tesoro
      else 
        return nil
      end
    end
    
    def to_s
      "#{@name}, Level: #{@level}"  
    end
    
    private
    
    def bringToLife
      @dead=false
    end

    def incrementLevels(l)
      if(@level + l <= @@MAXLEVEL)
        @level +=l
      else
        @level = @@MAXLEVEL
      end
    end
  
    def decrementLevels(l)
      if(@level-l >=1)
        @level-=l
      else
        @level = 1
      end  
    end
  
    def setPendingBadConsequence(b)
      @pendingBadConsequence=b
    end
  
    def applyPrize(m)
      nLevels = m.getLevelsGained
      incrementLevels(nLevels)
      nTreasures = m.getTreasuresGained
      if(nTreasures > 0)
        dealer = CardDealer.getInstance
        nTreasures.times do
          treasure = dealer.nextTreasure
          @hiddenTreasures << treasure
        end
      end
    end
  
    def applyBadConsequence(m)
      badConsequence = m.getBadConsequence
      nLevels = badConsequence.getLevels
      decrementLevels(nLevels)
      pendingBad = badConsequence.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
      setPendingBadConsequence(pendingBad)
    end
  
    def canMakeTreasureVisible(t)
      can_make = true
      
      if(t.getType == NapakalakiGame::TreasureKind::BOTHHANDS)
        @visibleTreasures.each do |i|
          if((i.getType == t.getType || i.getType == NapakalakiGame::TreasureKind::ONEHAND) && can_make)
            can_make = false
          end
        end
      else if(t.getType == NapakalakiGame::TreasureKind::ONEHAND && can_make)
          contador=0
          @visibleTreasures.each do |i|
            if(i.getType == NapakalakiGame::TreasureKind::BOTHHANDS)
              can_make = false
            end
            if(i.getType == t.getType && can_make)
              contador=contador+1
              if(contador==2)
                can_make = false             
              end
            end
          end
        else if(can_make)
            @visibleTreasures.each do |i|
              if(i.getType == t.getType && can_make)
                can_make=false
              end
            end
          end
        end
      end
      return can_make
    end
      
  
    def howManyVisibleTreasures(tKind)
      contador=0
      @visibleTreasures.each do |i|
        if(i.getType == tKind)
          contador+=1
        end  
      end
      return contador
    end
  
    def dieIfNoTreasures
      if(@visibleTreasures == 0 && @hiddenTreasures == 0)
        @dead = true
      end
    end
  
    def haveStolen 
      @canISteal=false
    end
    
    protected
    
    def getOponentLevel(m)
      return m.getCombatLevel
    end
    
    def shouldConvert
      d = Dice.instance
      aux = d.nextNumber
      return (aux == 1)      
    end
    
    def getCombatLevel
      bonus=0;      
      @visibleTreasures.each do |i|
        bonus+=i.getBonus
      end
      combatLevel=@level+bonus
      return combatLevel
    end
  
    #Borrar luego, lo dejo para que main sirva
    attr_reader:name
    attr_reader:hiddenTreasures
    attr_reader:visibleTreasures
    attr_reader:level
  end
end

File: PDOO/Napakalaki/src/NapakalakiGame/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Random;
import GUI.*;

/**
 *
 * @author nestor
 */
public class Player {

    static final int MAXLEVEL = 10;
    private String name;
    private int level;
    private boolean dead;
    private boolean canISteal;

    //Relaciones:
    protected Player enemy;
    private ArrayList<Treasure> hiddenTreasures;
    private ArrayList<Treasure> visibleTreasures;
    private BadConsequence pendingBadConsequence;

    public Player(String name) {
        this.name = name;
        this.level = 1;
        this.dead = true;
        this.canISteal = true;
        this.enemy = null;
        this.hiddenTreasures = new ArrayList();
        this.visibleTreasures = new ArrayList();
        this.pendingBadConsequence = null;
    }

    public void copyPlayer(Player p) {
        this.name = p.name;
        this.level = p.level;
        this.dead = p.dead;
        this.canISteal = p.canISteal;
        this.enemy = p.enemy;
        this.hiddenTreasures = p.hiddenTreasures;
        this.visibleTreasures = p.visibleTreasures;
        this.pendingBadConsequence = p.pendingBadConsequence;
    }

    protected int getOponentLevel(Monster m) {
        return m.getCombatLevel();
    }

    protected boolean shouldConvert() {
        Dice d = Dice.getInstance();
        int aux = d.nextNumber();
        return (aux == 1);
    }

    public String getName() {
        return name;
    }

    private void bringToLife() {
        dead = false;
    }

    public int getCombatLevel() {
        int total = 0;
        int bonus = 0;

        for (int i = 0; i < visibleTreasures.size(); i++) {
            bonus = bonus + visibleTreasures.get(i).getBonus();
        }
        total = level + bonus;
        return total;
    }

    public Player getEnemy() {
        return enemy;
    }

    private void incrementLevels(int i) {
        if (level + i <= MAXLEVEL) {
            level = level + i;
        } else {
            level = MAXLEVEL;
        }
    }

    private void decrementLevels(int i) {
        if (level - i >= 1) {
            level = level - i;
        } else {
            level = 1;
        }
    }

    private void setPendingBadConsequence(BadConsequence b) {
        pendingBadConsequence = b;
    }

    private void applyPrize(Monster m) {
        int nLevels = m.getLevelsGained();
        this.incrementLevels(nLevels);
        int nTreasures = m.getTreasuresGained();
        if (nTreasures > 0) {
            CardDealer dealer = CardDealer.getInstance();
            Treasure treasure = null;
            for (int i = 0; i < nTreasures; i++) {
                treasure = dealer.nextTreasure();
                this.hiddenTreasures.add(treasure);
            }
        }
    }

    private void applyBadConsequence(Monster m) {
        BadConsequence badConsequence = m.getBadConsequence();
        int nLevels = badConsequence.getLevels();
        this.decrementLevels(nLevels);
        BadConsequence pendingBad = badConsequence.adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);
        this.setPendingBadConsequence(pendingBad);

    }

    private boolean canMakeTreasureVisible(Treasure t) {
        boolean can_make = true;
        if (t.getType() == TreasureKind.BOTHHANDS) {
            for (int i = 0; i < visibleTreasures.size(); i++) {
                if ((visibleTreasures.get(i).getType() == t.getType() || visibleTreasures.get(i).getType() == TreasureKind.ONEHAND) && can_make) {
                    can_make = false;
                }
            }
        } else if (t.getType() == TreasureKind.ONEHAND && can_make) {
            int contador = 0;
            for (int i = 0; i < visibleTreasures.size(); i++) {
                if (visibleTreasures.get(i).getType() == TreasureKind.BOTHHANDS) {
                    can_make = false;
                }
                if (visibleTreasures.get(i).getType() == t.getType() && can_make) {
                    contador++;
                    if (contador == 2) {
                        can_make = false;
                    }
                }
            }

        } else if (can_make) {
            for (int i = 0; i < visibleTreasures.size(); i++) {
                if (visibleTreasures.get(i).getType() == t.getType() && can_make) {
                    can_make = false;
                }
            }
        }

        return can_make;
    }

    private int howManyTreasureVisible(TreasureKind tKind) {
        int contador = 0;

        for (int i = 0; i < visibleTreasures.size(); i++) {
            if (visibleTreasures.get(i).getType() == tKind) {
                contador++;
            }
        }

        return contador;
    }

    private void dieIfNoTreasures() {
        if (visibleTreasures.size() == 0 && hiddenTreasures.size() == 0) {
            dead = true;
        }
    }

    public boolean isDead() {
        return dead;
    }

    public ArrayList<Treasure> getHiddenTreasures() {
        return hiddenTreasures;
    }

    public ArrayList<Treasure> getVisibleTreasures() {
        return visibleTreasures;
    }
    
    public BadConsequence getPendingBadConsequence(){
        return pendingBadConsequence;
    }

    public CombatResult combat(Monster m) {
        CombatResult combatResult = null;
        int myLevel = this.getCombatLevel();
        int monsterLevel = this.getOponentLevel(m);
        if (myLevel > monsterLevel) {
            this.applyPrize(m);
            if (this.level >= MAXLEVEL) {
                combatResult = CombatResult.WINGAME;
            } else {
                combatResult = CombatResult.WIN;
            }
        } else {
            this.applyBadConsequence(m);
            if (this.shouldConvert()) {
                combatResult = CombatResult.LOSEANDCONVERT;
            } else {
                combatResult = CombatResult.LOSE;
            }
        }
        return combatResult;
    }

    public void makeTreasureVisible(Treasure t) {
        boolean canI = this.canMakeTreasureVisible(t);
        if (canI) {
            this.visibleTreasures.add(t);
            this.hiddenTreasures.remove(t);
        }

    }

    public void discardVisibleTreasure(Treasure t) {
        this.visibleTreasures.remove(t);
        if (pendingBadConsequence != null && !pendingBadConsequence.isEmpty()) {
            this.pendingBadConsequence.substractVisibleTreasure(t);
        }
        this.dieIfNoTreasures();
    }

    public void discardHiddenTreasure(Treasure t) {
        this.hiddenTreasures.remove(t);
        if (pendingBadConsequence != null && !pendingBadConsequence.isEmpty()) {
            this.pendingBadConsequence.substractHiddenTreasure(t);
        }
        this.dieIfNoTreasures();
    }

    public boolean validState() {
        boolean state;
        if ((pendingBadConsequence == null || pendingBadConsequence.isEmpty()) && hiddenTreasures.size() <= 4) {
            state = true;
        } else {
            state = false;
        }
        return state;
    }

    public void initTreasures() {
        CardDealer dealer = CardDealer.getInstance();
        Dice dice = Dice.getInstance();
        this.bringToLife();
        Treasure treasure = dealer.nextTreasure();
        this.hiddenTreasures.add(treasure);
        int number = dice.nextNumber();
        if (number > 1) {
            treasure = dealer.nextTreasure();
            this.hiddenTreasures.add(treasure);
        }
        if (number == 6) {
            treasure = dealer.nextTreasure();
            this.hiddenTreasures.add(treasure);
        }

    }

    public int getLevels() {
        return level;
    }

    public Treasure stealTreasure() {
        Treasure treasure = null;
        boolean canI = this.canISteal();
        if (canI) {
            boolean canYou = this.enemy.canYouGiveMeATreasure();
            if (canYou) {
                treasure = this.enemy.giveMeATreasure();
                this.hiddenTreasures.add(treasure);
                this.haveStolen();
            }
        }
        return treasure;
    }

    public void setEnemy(Player enemy) {
        this.enemy = enemy;
    }

    private Treasure giveMeATreasure() {
        boolean puede = canYouGiveMeATreasure();
        Treasure t;
        if (puede) {
            Random r = new Random();
            int aleatorio = 0;
            aleatorio = r.nextInt(hiddenTreasures.size());

            t = hiddenTreasures.get(aleatorio);
            hiddenTreasures.remove(aleatorio);
        } else {
            t = null;
        }
        return t;
    }

    public boolean canISteal() {
        return canISteal;
    }

    private boolean canYouGiveMeATreasure() {
        boolean puede;
        puede = hiddenTreasures.size() > 0;
        return puede;
    }

    private void haveStolen() {
        canISteal = false;
    }

    public void discardAllTreasures() {
        ArrayList<Treasure> copiavisible = new ArrayList(visibleTreasures);
        ArrayList<Treasure> copiaoculta = new ArrayList(hiddenTreasures);
        Iterator<Treasure> vit = copiavisible.iterator();
        Iterator<Treasure> hit = copiaoculta.iterator();
        while (vit.hasNext()) {
            Treasure treasure = vit.next();
            this.discardVisibleTreasure(treasure);
        }
        while (hit.hasNext()) {
            Treasure treasure = hit.next();
            this.discardHiddenTreasure(treasure);
        }
    }

    public String toString() {
        return name + ", Level: " + level;
    }
}
